### ISSUE TITLE
Support "Extract Method" | "Extract Variable" | "Move Symbol"

### ISSUE THREAD

**Issue by KotlinIsland** [DetachHead/basedpyright#223]
Sometimes my method/variable is 'inserted', and I would like it to be 'extracted', so I would appreciate these features.

Maybe we could use `rope` to extract the variables????

**Comment by DetachHead** [comment]
what exactly do each of these do

**Comment by KotlinIsland** [comment]
https://code.visualstudio.com/docs/editor/refactoring

**Comment by PhenomenaPh** [comment]
Yes, yes, yes! Huge support for refactoring tools. Without them, it's hard to use basedpyright 









Ask ChatGPT

**Comment by PhenomenaPh** [comment]
Yes, yes, yes! Huge support for refactoring tools. Without them, it's hard to use basedpyright

**Comment by dantetemplar** [comment]
Upvote

**Comment by DetachHead** [comment]
no need to comment "upvote", just click the 👍 emoji on the issue to upvote it

**Comment by iiian** [comment]
I know it's a tangent but if you'll entertain me asking, what is the official relationship between `basedpyright` and `Pyright Windsurf`? The [Pyright Windsurf extension](https://marketplace.windsurf.com/extension/Codeium/windsurfPyright) currently marks **this** repository as its home.

**Comment by DetachHead** [comment]
windsurf pyright is their fork of basedpyright. from their readme it looks like most of their changes are just changing the default settings to match pylance. windsurf is not associated with basedpyright in any way, but they are free to fork it, i have no issue with that. cursor did the same thing

**Comment by hnykda** [comment]
Is there any update on this? It's by far the most wanted feature by votes.

A lot of users of cursor are after this, as cursor is using basedpyright to ship their version of python experience. E.g. discussed here: https://forum.cursor.com/t/no-more-refactor-move-to-file-in-python/104103/31

I really appreciate basedpyright, thanks for that!

**Comment by DetachHead** [comment]
lately i've been trying to focus on bug fixes rather than new features because i think stability is very important, but this will probably be the next feature i try to implement considering how many upvotes it has

**Comment by emredo** [comment]
Hi guys, 'remove all unused imports' quick action option has disappeared, only 'remove unused import' option is visible. Do you think that it is related with this issue?

**Comment by DetachHead** [comment]
are you sure that code action came from pyright and not another tool like ruff? i don't think pyright ever had a code action to remove all unused imports

**Comment by emredo** [comment]
> are you sure that code action came from pyright and not another tool like ruff? i don't think pyright ever had a code action to remove all unused imports

Oh, thanks for your answer. I have double checked and noticed that these option comes from ruff. Sorry for carelessness

**Comment by DetachHead** [comment]
No worries, some editors make it difficult to tell which language server a diagnostic or code action  is coming from

**Comment by Mcklmo** [comment]
How come that support for this has stopped? I was using cursor for quite some time until this stopped working. Why not just use what works in VS Code? Why do you need to mark it a "feature", if there is a working version of this? I'm not happy!

**Comment by DetachHead** [comment]
cursor used to use pylance, but microsoft bans pylance from being used outside of vscode, so they had to switch to basedpyright which has not yet implemented all of pylance's features. feel free to direct your frustrations to https://github.com/microsoft/pylance-release/issues

**Comment by huuvan023** [comment]
Same issue

**Comment by KimJangHoon0** [comment]
please

**Comment by asukaminato0721** [comment]
I’ve implemented extract method in another project, here for ref

https://github.com/facebook/pyrefly/commit/39ec3446b2f5eeebbef4c6d4d6a99377b6295aaf
